File: src/exchangebookgenerator.py
import numpy as np
from typing import List, Dict

class ExchangeBookGenerator:
    """Génère un carnet Level‑1 multi‑exchange cohérent avec un NBBO donné.

    Correctifs 2025‑06‑23
    --------------------
    • **Aucun volume nul ou négatif** : on limite le nombre d’exchanges actifs à ce
      que chaque venue reçoive **au moins 100 actions** (lot minimal affiché).
    • **NBBO toujours doté d’un volume ≥100**.
    • Table d’exchanges par défaut mise à jour (plus de « Other »).
    """

    TICK = 0.01  # mouvement de prix minimal (1 ¢)

    # ──────────────────────────────────────────────
    #  PARAMÈTRES PAR DÉFAUT (modifiable côté PriceModel)
    # ──────────────────────────────────────────────
    DEFAULT_SHARES: Dict[str, float] = {
        "NSDQ": 0.30,
        "ARCA": 0.18,
        "BATS": 0.10,
        "BATY": 0.06,
        "EDGA": 0.03,
        "EDGX": 0.07,
        "IEXG": 0.05,
        "AMEX": 0.03,
        "NYSE": 0.12,
        "PHLX": 0.01,
        "MEMX": 0.03,
        "MIAX": 0.01,
        "BOSX": 0.00,   # volumes marginaux, gardé pour la variété
        "LTSE": 0.01,
    }

    DEFAULT_OFFSETS: Dict[int, float] = {0: 0.35, 1: 0.30, -1: 0.20, 2: 0.10, -2: 0.05}

    def __init__(self,
                 shares: Dict[str, float] | None = None,
                 offsets_p: Dict[int, float] | None = None,
                 seed: int | None = None):
        # Part de marché par venue ➔ transforme en distribution
        self.shares = shares or self.DEFAULT_SHARES.copy()
        total_share = sum(self.shares.values())
        if total_share == 0:
            raise ValueError("Sum of market shares must be > 0")
        self.exchs   = np.array(list(self.shares.keys()))
        self.share_p = np.array(list(self.shares.values()), dtype=float) / total_share

        # Distribution d’offsets (en ticks)
        self.offsets_p = offsets_p or self.DEFAULT_OFFSETS.copy()
        self.kvals  = np.array(list(self.offsets_p.keys()), dtype=int)
        probs       = np.array(list(self.offsets_p.values()), dtype=float)
        self.offset_p = probs / probs.sum()

        self.rng = np.random.default_rng(seed)

    # ──────────────────────────────────────────────
    #  MÉTHODES INTERNES
    # ──────────────────────────────────────────────
    def _pick_exchanges(self, total_size: int) -> np.ndarray:
        """Sélectionne un nombre de venues compatible avec le volume total.

        • Minimum 3 venues.
        • Maximum : min(7, total_size // 100).  Chaque venue aura ≥100 shares.
        """
        max_n = min(7, max(1, total_size // 100))
        n = self.rng.integers(3, max_n + 1) if max_n >= 3 else 1
        return self.rng.choice(self.exchs, size=n, replace=False, p=self.share_p)

    def _dirichlet_sizes(self, total: int, n: int) -> np.ndarray:
        """Répartition de `total` en n blocs ≥100.

        Algorithme :
        1. Tire un Dirichlet `w`.
        2. Transforme en lots de 100 actions via floor.
        3. Répartit le reste (total - somme)  par incréments de 100 jusqu’à épuisement.
        """
        if n == 1:
            return np.array([total], dtype=int)

        # Étape 1‑2
        w = self.rng.dirichlet(np.ones(n))
        sizes = np.floor(w * total / 100).astype(int) * 100

        # Garantit un minimum de 100 pour chaque venue
        shortfall = (sizes == 0).sum() * 100
        sizes[sizes == 0] = 100

        # Ajuste le surplus/déficit par rapport à total
        diff = total - sizes.sum()

        # Cas surplus : retire 100 sur les plus grosses tailles
        while diff < 0:
            idx = sizes.argmax()
            if sizes[idx] > 100:
                sizes[idx] -= 100
                diff += 100
            else:
                break  # sécurité

        # Cas déficit : ajoute 100 aux venues au hasard
        while diff > 0:
            idx = self.rng.integers(n)
            sizes[idx] += 100
            diff -= 100

        return sizes

    def _draw_offsets(self, n: int, ensure_best: bool = False) -> np.ndarray:
        k = self.rng.choice(self.kvals, size=n, p=self.offset_p)
        if ensure_best and not np.any(k == 0):
            k[self.rng.integers(n)] = 0
        return k * self.TICK

    # ──────────────────────────────────────────────
    #  API PUBLIQUE
    # ──────────────────────────────────────────────
    def generate(self, tick: Dict) -> List[Dict]:
        """Produit les quotes par exchange pour un tick NBBO.

        `tick` doit contenir : `time`, `priceBid`, `priceAsk`, `sizeBid`, `sizeAsk`.
        """
        nbbo_bid, nbbo_ask = tick["priceBid"], tick["priceAsk"]
        size_bid, size_ask = tick["sizeBid"], tick["sizeAsk"]

        exchs = self._pick_exchanges(total_size=min(size_bid, size_ask))
        n     = len(exchs)

        bid_sizes = self._dirichlet_sizes(size_bid, n)
        ask_sizes = self._dirichlet_sizes(size_ask, n)

        bid_off   = self._draw_offsets(n, ensure_best=True)
        ask_off   = self._draw_offsets(n, ensure_best=True)

        # Garantit un volume sur la(les) ligne(s) NBBO
        nbbo_bid_idx = np.where(bid_off == 0)[0]
        if bid_sizes[nbbo_bid_idx].min() < 100:
            deficit = 100 - bid_sizes[nbbo_bid_idx].min()
            donor = bid_sizes.argmax()
            if bid_sizes[donor] - deficit >= 100:
                bid_sizes[donor] -= deficit
                bid_sizes[nbbo_bid_idx[0]] += deficit

        nbbo_ask_idx = np.where(ask_off == 0)[0]
        if ask_sizes[nbbo_ask_idx].min() < 100:
            deficit = 100 - ask_sizes[nbbo_ask_idx].min()
            donor = ask_sizes.argmax()
            if ask_sizes[donor] - deficit >= 100:
                ask_sizes[donor] -= deficit
                ask_sizes[nbbo_ask_idx[0]] += deficit

        # Construction finale
        quotes: List[Dict] = []
        for i, ex in enumerate(exchs):
            quotes.append({
                "time":     tick["time"],
                "exchange": ex,
                "priceBid": round(nbbo_bid + bid_off[i], 2),
                "sizeBid":  int(bid_sizes[i]),
                "priceAsk": round(nbbo_ask + ask_off[i], 2),
                "sizeAsk":  int(ask_sizes[i]),
            })
        return quotes

File: src/models/ExchangeBookGenerator.ts
// ExchangeBookGenerator.ts - Converted from exchangebookgenerator.py

export interface Quote {
  time: Date;
  exchange: string;
  priceBid: number;
  sizeBid: number;
  priceAsk: number;
  sizeAsk: number;
}

interface Tick {
  time: Date;
  priceBid: number;
  priceAsk: number;
  sizeBid: number;
  sizeAsk: number;
}

export class ExchangeBookGenerator {
  private static readonly TICK = 0.01;

  private static readonly DEFAULT_SHARES: Record<string, number> = {
    NSDQ: 0.30,
    ARCA: 0.18,
    BATS: 0.10,
    BATY: 0.06,
    EDGA: 0.03,
    EDGX: 0.07,
    IEXG: 0.05,
    AMEX: 0.03,
    NYSE: 0.12,
    PHLX: 0.01,
    MEMX: 0.03,
    MIAX: 0.01,
    BOSX: 0.00,
    LTSE: 0.01,
  };

  private static readonly DEFAULT_OFFSETS: Record<number, number> = {
    0: 0.35,
    1: 0.30,
    "-1": 0.20,
    2: 0.10,
    "-2": 0.05,
  };

  private exchs: string[];
  private shareP: number[];
  private kvals: number[];
  private offsetP: number[];
  private rng: () => number;

  constructor(
    shares?: Record<string, number>,
    offsetsP?: Record<number, number>,
    seed?: number
  ) {
    const sharesData = shares || ExchangeBookGenerator.DEFAULT_SHARES;
    const offsetsData = offsetsP || ExchangeBookGenerator.DEFAULT_OFFSETS;

    // Setup market shares
    const totalShare = Object.values(sharesData).reduce((a, b) => a + b, 0);
    if (totalShare === 0) {
      throw new Error("Sum of market shares must be > 0");
    }

    this.exchs = Object.keys(sharesData);
    this.shareP = Object.values(sharesData).map((v) => v / totalShare);

    // Setup offsets
    this.kvals = Object.keys(offsetsData).map(Number);
    const probs = Object.values(offsetsData);
    const totalProb = probs.reduce((a, b) => a + b, 0);
    this.offsetP = probs.map((p) => p / totalProb);

    // Simple seeded random (not cryptographically secure)
    if (seed !== undefined) {
      let s = seed;
      this.rng = () => {
        s = (s * 9301 + 49297) % 233280;
        return s / 233280;
      };
    } else {
      this.rng = Math.random;
    }
  }

  private pickExchanges(totalSize: number): string[] {
    const maxN = Math.min(7, Math.max(1, Math.floor(totalSize / 100)));
    const n = maxN >= 3 ? Math.floor(this.rng() * (maxN - 2)) + 3 : 1;
    return this.weightedSample(this.exchs, this.shareP, n, false);
  }

  private weightedSample(
    items: string[],
    weights: number[],
    n: number,
    replace: boolean
  ): string[] {
    const result: string[] = [];
    const available = [...items];
    const availableWeights = [...weights];

    for (let i = 0; i < n; i++) {
      const totalWeight = availableWeights.reduce((a, b) => a + b, 0);
      let random = this.rng() * totalWeight;

      for (let j = 0; j < available.length; j++) {
        random -= availableWeights[j];
        if (random <= 0) {
          result.push(available[j]);
          if (!replace) {
            available.splice(j, 1);
            availableWeights.splice(j, 1);
          }
          break;
        }
      }
    }

    return result;
  }

  private dirichletSizes(total: number, n: number): number[] {
    if (n === 1) {
      return [total];
    }

    // Simple Dirichlet approximation using gamma distribution
    const alpha = Array(n).fill(1);
    const gamma = alpha.map(() => {
      // Simple gamma(1,1) approximation
      return -Math.log(this.rng());
    });
    const sum = gamma.reduce((a, b) => a + b, 0);
    const w = gamma.map((g) => g / sum);

    let sizes = w.map((weight) => Math.floor((weight * total) / 100) * 100);

    // Ensure minimum 100 for each
    sizes = sizes.map((s) => (s === 0 ? 100 : s));

    // Adjust to match total
    let diff = total - sizes.reduce((a, b) => a + b, 0);

    while (diff < 0) {
      const maxIdx = sizes.indexOf(Math.max(...sizes));
      if (sizes[maxIdx] > 100) {
        sizes[maxIdx] -= 100;
        diff += 100;
      } else {
        break;
      }
    }

    while (diff > 0) {
      const idx = Math.floor(this.rng() * n);
      sizes[idx] += 100;
      diff -= 100;
    }

    return sizes;
  }

  private drawOffsets(n: number, ensureBest: boolean = false): number[] {
    const k = Array(n)
      .fill(0)
      .map(() => {
        const totalWeight = this.offsetP.reduce((a, b) => a + b, 0);
        let random = this.rng() * totalWeight;
        for (let i = 0; i < this.kvals.length; i++) {
          random -= this.offsetP[i];
          if (random <= 0) {
            return this.kvals[i];
          }
        }
        return this.kvals[0];
      });

    if (ensureBest && !k.includes(0)) {
      k[Math.floor(this.rng() * n)] = 0;
    }

    return k.map((val) => val * ExchangeBookGenerator.TICK);
  }

  generate(tick: Tick): Quote[] {
    const nbboBid = tick.priceBid;
    const nbboAsk = tick.priceAsk;
    const sizeBid = tick.sizeBid;
    const sizeAsk = tick.sizeAsk;

    const exchs = this.pickExchanges(Math.min(sizeBid, sizeAsk));
    const n = exchs.length;

    const bidSizes = this.dirichletSizes(sizeBid, n);
    const askSizes = this.dirichletSizes(sizeAsk, n);

    const bidOff = this.drawOffsets(n, true);
    const askOff = this.drawOffsets(n, true);

    // Ensure NBBO has volume
    const nbboBidIdx = bidOff.findIndex((o) => o === 0);
    if (nbboBidIdx !== -1 && bidSizes[nbboBidIdx] < 100) {
      const deficit = 100 - bidSizes[nbboBidIdx];
      const donor = bidSizes.indexOf(Math.max(...bidSizes));
      if (bidSizes[donor] - deficit >= 100) {
        bidSizes[donor] -= deficit;
        bidSizes[nbboBidIdx] += deficit;
      }
    }

    const nbboAskIdx = askOff.findIndex((o) => o === 0);
    if (nbboAskIdx !== -1 && askSizes[nbboAskIdx] < 100) {
      const deficit = 100 - askSizes[nbboAskIdx];
      const donor = askSizes.indexOf(Math.max(...askSizes));
      if (askSizes[donor] - deficit >= 100) {
        askSizes[donor] -= deficit;
        askSizes[nbboAskIdx] += deficit;
      }
    }

    // Build quotes
    const quotes: Quote[] = [];
    for (let i = 0; i < exchs.length; i++) {
      quotes.push({
        time: tick.time,
        exchange: exchs[i],
        priceBid: Math.round((nbboBid + bidOff[i]) * 100) / 100,
        sizeBid: Math.floor(bidSizes[i]),
        priceAsk: Math.round((nbboAsk + askOff[i]) * 100) / 100,
        sizeAsk: Math.floor(askSizes[i]),
      });
    }

    return quotes;
  }
}
